Split and Merge algorithm

AleCarminati opened the pull request:
Good afternoon,

we implemented the Split and Merge algorithm in bayesmix. 

We ran tests on galaxy and faithful datasets with the same settings that were used in the master's thesis on bayesmix and with the following parameters for the algorithm:
- 5 iterations of restricted Gibbs Sampling
- 1 iteration of MH sampling
- 1 iteration of full Gibbs Sampling, i.e. Neal3 

We obtained interesting results:
- on galaxy we obtained an ESS on the number of clusters of 622.256. The run time was 1.648s, while the run time for Neal2, Neal3 and Neal8 was, respectively, 0.577s, 0.828s, and 0.819s.
- on faithful we obtained an ESS on the number of clusters of 659.87. The run time was 20.568s, while the run time for Neal2, Neal3 and Neal8 was, respectively, 2.322s, 6.164s, and 8.093s.

We include the density and the cluster estimates we obtained with Split and Merge.

[galaxy_cluster](https://user-images.githubusercontent.com/48724629/152831669-076fcb22-209f-47a2-bfa1-b001c886c658.png)
[galaxy_density](https://user-images.githubusercontent.com/48724629/152831692-fa1a8a29-91b3-411a-95cb-f0c2787d4ce8.png)

[faithful_cluster](https://user-images.githubusercontent.com/48724629/152831717-910f59a9-480e-42dd-a5fc-5ca9ca629dbe.png)
[faithful_density](https://user-images.githubusercontent.com/48724629/152831725-fc63f257-594a-491a-9cc9-ab812c3b90cb.png)

mberaha commented:
I will have a look as soon as possible. In the meantime, what is the ESS per second for all the models? Can you compare that?

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+      j++;
+    }
+
+    compute_S(i,j);
+    compute_C_launch(i,j);
+
+    // Restricted GS steps
+    for(unsigned int t=0; t<T; ++t){
+      restricted_GS(i,j);
+    }
```
I would give more explicative names to all of these methods

AleCarminati replied:
Concerning the function `restricted_GS`, the problem is in the term "restricted Gibbs Sampling" or in the absence of information about what the function does?
More precisely, should we find a more explicative substitute for "restricted Gibbs Sampling" or should we enhance the name of the function to make it more explicative (e.g. something like `restricted_GS_step` or `sample_restricted_GS`)?

mberaha replied:
I think that restricted_gibbs_sampling would be fine, provided that you write a meaningful docstring about what this function is actually doing in the .h file

mberaha reviewed (commented):
Please run the clang formatter as explained in https://github.com/bayesmix-dev/bayesmix/blob/master/CONTRIBUTORS.md since at the moment the GitHub diff is almost impossible to read!

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+}
+
+void SplitAndMergeAlgorithm::compute_S(const unsigned int i, const unsigned int j) {
+    unsigned int lengthS=0;
+    for(int k=0; k<allocations.size(); k++){
+        if(allocations[k]==allocations[i]||allocations[k]==allocations[j])
+            lengthS++;
+    }
+    lengthS=lengthS-2;
```
I think that the same can be achieved with

```
int num_obs_state = 0;
if (allocations[i] == allocations[j] ) {
   num_obs_state = unique_values[allocations[i]]->get_card();
} else {
  num_obs_state = unique_values[allocations[i]]->get_card() + 
    unique_values[allocations[j]]->get_card();
}
num_obs_state = num_obs_state - 2;
```

mberaha replied:
which avoids the for loop

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+            lengthS++;
+    }
+    lengthS=lengthS-2;
+    S={}; //Necessary, since it needs to be zero at each iter, not only at the beginning
+    S.resize(lengthS,0);
+    unsigned int index=0;
+    for (int k = 0; k < allocations.size(); ++k) {
+        if ((allocations[k]==allocations[i]||allocations[k]==allocations[j])&& (k!=i) && (k!=j)) {
+            S[index]=k;
+            index++;
+        }
+    }
```
Actually you don't even need the above code, you cold just write

```
S = {};
if (allocations[i] == allocations[j] ) {
   S = unique_values[allocations[I]]->get_data_idx();
} else {
   const auto& first_clus = unique_values[allocations[I]]->get_data_idx();
   const auto& second_clus = unique_values[allocations[j]]->get_data_idx();
   std::set_union(first_clus.begin(), first_clus.end(),
          second_clus.begin(), second_clus.end(),
           std::inserter(S, S.begin()));
}
S.erase(allocations[I]);
S.erase(allocations[j]);
```

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+  cl[1]->add_datum(j, data.row(j), true);
+}
+
+void SplitAndMergeAlgorithm::split_or_merge(const unsigned int i, const unsigned int j){
```
I think this method be split in several functions, for instance

```
if (allocations[i]==allocations[j]) {
  split();
else {
   merge()
};
```

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+    unsigned int z=0;
+    std::set<int> set_i=cl[0]->get_data_idx();
+    std::set<int> set_j=cl[1]->get_data_idx();
+    std::set<int>::const_iterator iter_i=set_i.cbegin();
```
I don't think you actually need an iterator, since you're only using *iter in the rest of the code. It's cleaner to let these `int i_idx; j_idx`

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+    for(unsigned int k=0; k < clSplit.size(); k++){
+      if(((z<S.size())and(k==S[z]))or((k==i)or(k==j))){ //since S doesn't contain i and j but we put them in data_i or data_j
+        if(*iter_i==k){
+          if (!data_i->get_card()){ //first iteration
+            p_i+=data_i->prior_pred_lpdf(data.row(k));
+          }else{
+            p_i+=data_i->conditional_pred_lpdf(data.row(k));
+          }
```
I think we should make it so there is no need for this check, i.e. `conditional_pred_lpdf` returns the `prior_pred_lpdf` if the cluster is empty. What do you think @brunoguindani @TeoGiane ?

brunoguindani replied:
I wholeheartedly agree!

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+    const double alpha=mixing->get_mass_new_cluster(allocations.size(), false,
+                                       true, LabI+1); //LabI should be number of clusters - 1, alternatively unique_values.size()
+    // data_i->get_card()-1 = n. of data with LabelI different from i, data_j->get_card()-1= n. of data with LabelJ different from j (+1 because we are using gamma function)
+    const double p2=tgamma(data_j->get_card()-1-1+1)*tgamma(data_i->get_card()-1-1+1)/(S.size()+2-1)*alpha;
+    const double p3=std::exp(p_i+p_j-p_J); 
+    const double AcRa=std::min(1.0,p1*p2*p3); //acceptance ratio 
```
All of these computations should be better performed in log-scale. 
You have the `lgamma` function that computes a numerically stable version of the log of the gamma function.
Also the `-1 + 1 +1` kind of weirds me out

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+    const double p3=std::exp(p_i+p_j-p_J); 
+    const double AcRa=std::min(1.0,p1*p2*p3); //acceptance ratio 
+    if(accepted_proposal(AcRa)){
+      unique_values.push_back(unique_values[0]->clone());
```
I would move all of the code in this `if` in a separate method

AleCarminati replied:
We moved all the code in the `if` in an auxiliary function, called `proposal_update_allocations`, which can be used both in the split and in the merge case. The function knows in which case it is working thanks to the boolean parameter `split`.
We decided to use a single function for both cases because the structure is exactly the same, the only things that change are the data indexes of the points that must be moved and the destination cluster.

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+      }
+    }
+  }else{
+    std::vector<unsigned int> clMerge (allocations.size()); 
```
new method here

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,408 @@
+    }
+        
+    const double p1=q;
+    const double alpha=mixing->get_mass_new_cluster(allocations.size(), false,
+                                       true,unique_values.size());
+    // data_i->get_card()-1 = n. of data with LabelI different from i, data_j->get_card()-1= n. of data with LabelJ different from j (+1 because we are using gamma function)
+    const double p2=tgamma(data_i->get_card()-1-1+1)*tgamma(data_j->get_card()-1-1+1)/(S.size()+2-1)*alpha;
```
Can you explain what is actually going on here? It seems to me that you're computing the prior probability of a given clustering, is it true? In that case, these numbers are specific to the DP prior, but we should be able to generalise to different mixings

AleCarminati replied:
The variable p2 represents the ratio between the prior probability of the proposal cluster configuration (i.e. the split or merge configuration) and the prior probability of the current configuration. 
For the split case, the formula is:
![Cattura](https://user-images.githubusercontent.com/48724629/153051274-124da83e-bfee-4db5-93ed-7825faf0b73a.PNG)
which is used in line [183](https://github.com/vnardi/bayesmix/blob/b696efa5417915702f561fa4781dde6b7d6f1989/src/algorithms/split_and_merge_algorithm.cc#L183).
For the merge case there's a different formula w.r.t. the one we wrote in line 301, we'll check if indeed we made an error.
Anyway, the formulas are equations (3.4) and (3.5) of the paper by Jain and Neal (2004).

Given that the formula is obtained by the computation of masses of clusters, we'll work to write them in a more general way by using get_mass_new_cluster and get_mass_existing cluster.

On a side note, in the [initialization](https://github.com/vnardi/bayesmix/blob/b696efa5417915702f561fa4781dde6b7d6f1989/src/algorithms/split_and_merge_algorithm.cc#L19-L23) of the algorithm, we decided to make sure that the mixing used in the algorithm is a DP because the article by Jain and Neal studied the behavior only with that particular mixing.
Given that we are generalizing the code to work also with other mixings, should we remove this control or should we keep it until we find academic results about the correct functioning of the algorithm with a generic mixing?

mberaha replied:
> The variable p2 represents the ratio between the prior probability of the proposal cluster configuration (i.e. the split or merge configuration) and the prior probability of the current configuration. For the split case, the formula is: ![Cattura](https://user-images.githubusercontent.com/48724629/153051274-124da83e-bfee-4db5-93ed-7825faf0b73a.PNG) which is used in line [183](https://github.com/vnardi/bayesmix/blob/b696efa5417915702f561fa4781dde6b7d6f1989/src/algorithms/split_and_merge_algorithm.cc#L183). For the merge case there's a different formula w.r.t. the one we wrote in line 301, we'll check if indeed we made an error. Anyway, the formulas are equations (3.4) and (3.5) of the paper by Jain and Neal (2004).
> 
> Given that the formula is obtained by the computation of masses of clusters, we'll work to write them in a more general way by using get_mass_new_cluster and get_mass_existing cluster.

Ok great, I think this should be taken care of in the `Mixing` object and there is no easy way to do it at the moment. Please leave a comment in the code with what exactly is this `p2` (also it should be computed in log-scale using the `lgamma` function and called `log_prior_proba`). And add
```
// FIXME@Mario: this needs generalization to other Mixings!
```
at the end of the comment block.

AleCarminati replied:
The computation of `p2` (now called `log_ratio_prior_prob`) has been moved to an auxiliary function, called `compute_log_ratio_like_and_prior`. In the split case, this function populates three `Hierarchy` objects: two for the two split proposal clusters (c_i^{split} and c_j^{split}) and one for the original cluster (c_i). Before adding each datum, we update `log_ratio_prior_prob` with `get_mass_existing_cluster()`. In this way, when a DP is used, we obtain exactly the formula we reported in the last comment, but it is also possible to generalize the computation to other mixings (until this computations will be added to the `Mixing` object). 
The merge case is rather similar in the computations.

mberaha replied:
Yes I'm actually sure that you can use split and merge with any mixing as long as you have an EPPF, several papers do so... You can remove the check of mixing being a DP!

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,462 @@
+    }
+  }
+  /* We update the posterior parameters only at the last insertion to ease
+   * computations.
```
+1

AleCarminati commented:
> I will have a look as soon as possible. In the meantime, what is the ESS per second for all the models? Can you compare that? I would also like to see the ESS/s when varying the parameters of the split-merge algorithm (number of restricted scans, number of metropolis-hastings moves).

As requested, we computed the ESS/s of our algorithm (using the code in commit fec8db5) with different parameters and we compared it with the other marginal algorithms in the library.
Here are the results:
[galaxy_ess_s](https://user-images.githubusercontent.com/48724629/153567771-e3ce742f-99fc-4c08-bd69-1c5443579dca.png)
[faithful_ess_s](https://user-images.githubusercontent.com/48724629/153567789-83766332-bdc2-4f67-94b0-cf1d543d1811.png)

We saw that, in general, the value of the ESS for Split and Merge is higher than the value for the other algorithms, but the run time is also higher. This makes our algorithm worth using (from the point of view of ESS/s) only with certain combinations of parameters. 
We analyzed Neal2, which has an ESS/s much higher that all the other algorithms: we found out that this extreme value is due to a very low run time, which compensates the ESS, lower than the one of all the other algorithms.

mberaha commented:
> As requested, we computed the ESS/s of our algorithm (using the code in commit [fec8db5](https://github.com/bayesmix-dev/bayesmix/commit/fec8db54494b747eed5695ea43d04df82fa484b3)) with different parameters and we compared it with the other marginal algorithms in the library. Here are the results: [galaxy_ess_s](https://user-images.githubusercontent.com/48724629/153567771-e3ce742f-99fc-4c08-bd69-1c5443579dca.png) [faithful_ess_s](https://user-images.githubusercontent.com/48724629/153567789-83766332-bdc2-4f67-94b0-cf1d543d1811.png)
> 
> We saw that, in general, the value of the ESS for Split and Merge is higher than the value for the other algorithms, but the run time is also higher. This makes our algorithm worth using (from the point of view of ESS/s) only with certain combinations of parameters. We analyzed Neal2, which has an ESS/s much higher that all the other algorithms: we found out that this extreme value is due to a very low run time, which compensates the ESS, lower than the one of all the other algorithms.

Can you try with a few other simulated datasets as the dimension increase? For instance you simulate data from a 2 component mixture, the first has mean equal to -3/sqrt(d) and the second one to 3/sqrt(d) (these are d-dimensional vectors with all entries equal) and identity covariance. You could compare the various algorithms for d=2,5,10,25. Set the number of initial clusters to 1. In my experience neal3 should be stuck to always finding only 1 cluster, while split-merge might find two. This could give a major advantage to split-merge

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,462 @@
+
+void SplitAndMergeAlgorithm::split(const unsigned int first_random_idx,
+                                   const unsigned int second_random_idx) {
+  double log_ratio_transition_prob = -restricted_gibbs_sampling(true);
```
I think that this call can be avoided since you call `split` after you have already called  `restricted_gibbs_sampling`

mberaha commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,462 @@
+    restricted_gs_unique_values[allocations_restricted_gs[k]]->remove_datum(
+        restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+        update_hierarchy_params());
+
+    Eigen::VectorXd logprobas(2);
+
+    if (restricted_gs_unique_values[0]->get_card() >= 1 and
+        restricted_gs_unique_values[1]->get_card() >= 1) {
+      logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+      logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+    } else {
+      if (restricted_gs_unique_values[1]->get_card() == 0) {
+        logprobas(1) = restricted_gs_unique_values[1]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      } else {
+        logprobas(0) = restricted_gs_unique_values[0]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      }
+    }
+    if (allocations[restricted_gs_data_idx[k]] ==
+        allocations[first_random_idx]) {
+      allocations_restricted_gs[k] = 0;
+      restricted_gs_unique_values[0]->add_datum(
+          restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+          update_hierarchy_params());
+      log_ratio_transition_prob += log(stan::math::softmax(logprobas)(0));
+    } else {
+      allocations_restricted_gs[k] = 1;
+      restricted_gs_unique_values[1]->add_datum(
+          restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+          update_hierarchy_params());
+      log_ratio_transition_prob += log(stan::math::softmax(logprobas)(1));
+    }
+  }
```
It looks to me like this for loop is very similar to the one in `restricted_gibbs_sampling`, am I right? Isn't it possible to avoid redoing this computation and saving all these stuff when you call `restricted_gibbs_sampling()`?

AleCarminati replied:
Yes, this computation is very similar to `restricted_gibbs_sampling`, with the unique difference that, for each point, the new cluster is not sampled, but it is the cluster of the original configuration.
This step must be done after the restricted Gibbs sampling cycles, so it can't be done during them.

From the point of view of the code, we added a parameter to `restricted_gibbs_sampling` in order to do the computations in the lines above: in this way in `merge` we just need to do a call to `restricted_gibbs_sampling`.
After this change, observing the functions `split` and `merge` we noted that they are nearly identical. 
Is it ok if we remove these two functions and move their code inside `split_or_merge`, using a parameter to identify if we are in the split or in the merge case? In this way, we could avoid duplicated code.

mberaha replied:
That would be great

mberaha reviewed (commented):
The code looks good to me! I have left some comments here and there. The major one is about the possibility of saving all important stuff when calling `restricted_gibbs_sampling` instead of computing it in the `merge` method. What do you think?

AleCarminati commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,462 @@
+
+    Eigen::VectorXd logprobas(2);
+
+    if (restricted_gs_unique_values[0]->get_card() >= 1 and
+        restricted_gs_unique_values[1]->get_card() >= 1) {
+      logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+      logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+    } else {
+      if (restricted_gs_unique_values[1]->get_card() == 0) {
+        logprobas(1) = restricted_gs_unique_values[1]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      } else {
+        logprobas(0) = restricted_gs_unique_values[0]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      }
+    }
```
By looking at the code we found out that, in this case, the lines 285-296 (the `else` branch) will never be executed, because `restricted_gs_unique_values[0]` will always contain the point `first_random_idx` and `restricted_gs_unique_values[1]` will always contain the point `second_random_idx`, because in the article on Split and Merge it is explicitly said these two points must not be considered when we do the restricted Gibbs Sampling, so they will never be moved from their initial position.
Therefore, we will proceed in removing these lines, so we can't put the `//FIXME` there.

Is it ok if we put it in line [177](https://github.com/vnardi/bayesmix/blob/fec8db54494b747eed5695ea43d04df82fa484b3/src/algorithms/split_and_merge_algorithm.cc#L177), where there's a if/else with a similar meaning?

mberaha replied:
sure, thanks

brunoguindani commented:
Hello, I will review this PR soon. In the meantime, could you please mark conversations in this thread as resolved if the proposed changes have already been implemented? Thank you!

brunoguindani commented on algo_marg_settings.asciipb:
```diff
@@ -25,3 +25,12 @@ init_num_clusters: 3
+splitmerge_n_mh_updates: 1
+
+# Split and Merge number of full GS scans for each iteration
+splitmerge_n_full_gs_updates: 1
```
This missing EOF newline, as well as other stuff in this PR, tells me you have not installed the pre-commit hook. Please follow instructions in [CONTRIBUTORS.md](https://github.com/bayesmix-dev/bayesmix/blob/master/CONTRIBUTORS.md) so that stuff like this is automatically taken care of whenever you create a new commit (you may need to re-add modified files to the commit before pushing). Actually I'm not sure if the hook only takes care of C++ files. Therefore, please check that this particular newline is there before pushing, and fix it manually if needed

AleCarminati replied:
We installed the pre-commit hooks starting from commit [fec8db5](https://github.com/bayesmix-dev/bayesmix/pull/116/commits/fec8db54494b747eed5695ea43d04df82fa484b3), indeed the code in C++ is formatted differently w.r.t. how we wrote it. 

We will fix the newline. Are there other formatting errors that we can fix manually?

brunoguindani replied:
I could have sworn I had seen a line with over 79 characters, but now it's nowhere to be found. My bad

mberaha replied:
"don't drink and revise"!

brunoguindani commented on algo_marg_settings.asciipb:
```diff
@@ -25,3 +25,12 @@ init_num_clusters: 3
 #  know what you're doing.)
 neal8_n_aux: 3
+
+# Split and Merge number of restricted GS scans for each MH step.
```
I'd put a title to make it clear that the following parameters are only used with the S&M algorithm

brunoguindani commented on src/algorithms/split_and_merge_algorithm.h:
```diff
@@ -0,0 +1,149 @@
+ protected:
+  void print_startup_message() const override;
+
+  /* We need to update the parameters when we add or remove a datum from a
+   * cluster to esure that conditional_pred_lpdf returns a correct value.
+   */
+  bool update_hierarchy_params() override { return true; }
```
I don't think this comment is needed, since it's a purely technical detail

brunoguindani commented on src/algorithms/split_and_merge_algorithm.h:
```diff
@@ -0,0 +1,149 @@
+
+  void full_gibbs_sampling();
+
+  /* After picking at random two points, the Split and Merge algorithm
+   * takes all the points that are in the same clusters as the first two and
+   * assigns to each of them, at random, one of two temporary clusters. Then,
+   * it updates these assignments by doing a restricted Gibbs sampling,
+   * restricted in the sense that it considers only the points cited above
+   * and only the two temporary clusters.
+   * The result of this operation is used to compute the acceptance
+   * probability for the split or merge proposal.
+   * For more information, check the article on Split and Merge by
+   * Jain and Neal (2004).
+   *
+   * This function computes one iteration of the restricted Gibbs sampling.
+   *
+   * If return_log_res_prod is true, the function returns the log of the
+   * probability of the transition that restricted_gs_unique_values has done
+   * in the function.
+   * If return_log_res_prod is false, ignore the return value.
+   *
+   * If step_to_original_clust is true, the function moves all the points
+   * in restricted_gs_data_idx to their original clustering configuration.
+   * This step is used in the merge case to compute the transition probability
+   * from the restricted gibbs sampling state to the original split
+   * configuration.
+   */
+  double restricted_gibbs_sampling(const double first_random_idx,
+                                   bool return_log_res_prod = false,
+                                   bool step_to_original_clust = false);
```
I'd rather have the big explanation on how the algorithm works at the beginning of the file. Method docstrings should only contain implementation details or a brief summary on what the method does

brunoguindani commented on src/algorithms/split_and_merge_algorithm.h:
```diff
@@ -0,0 +1,149 @@
+
+  /* Vector that contains the indexes of the data points that are considered
+   * in the restricted Gibbs sampling.
+   *
+   * For more information about restricted Gibbs Sampling, check out the
+   * comment on function restricted_gibbs_sampling().
```
This bit is unncesessary given that the names are self-explicative as is. Same for the other 3 instances of this sentence below

brunoguindani commented on src/algorithms/split_and_merge_algorithm.h:
```diff
@@ -0,0 +1,149 @@
+   * The parameter "split" represents if the function is called in the "split"
+   * case or in the "merge" case.
+   */
+  void compute_log_ratio_like_and_prior(const unsigned int first_random_idx,
+                                        const unsigned int second_random_idx,
+                                        bool split,
+                                        double &log_ratio_prior_prob,
+                                        double &log_ratio_likelihoods);
```
I think this would be cleaner if you returned log ratios as an `std::pair<double>` rather than using references, but I'm curious on what @mberaha and @TeoGiane think about this. Also, this function name is way too long; maybe `compute_log_ratios()`?

mberaha replied:
Completely agree, you could have

```
std::pair<double>  compute_log_ratios(...);
```
and when you call it

```
auto [log_lik_ratio, log_prior_ratio] = complute_log_ratios(...);
```

brunoguindani commented on src/algorithms/split_and_merge_algorithm.h:
```diff
@@ -0,0 +1,149 @@
+                                   const unsigned int second_random_idx,
+                                   const bool split);
+
+  void full_gibbs_sampling();
```
Consider replacing `gibbs_sampling` with `gs` in both this name and the following one. You are already using the abbreviation everywhere else anyways. Regardless, I would make it explicit in the introductory text at the beginning of the file for clarity's sake: "...Gibbs Sampling (GS) ..."

brunoguindani commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,393 @@
+    unsigned int first_random_idx = bayesmix::categorical_rng(probas, rng, 0);
+    probas = Eigen::VectorXd::Constant(n_data - 1, 1.0 / (n_data - 1));
+    unsigned int second_random_idx = bayesmix::categorical_rng(probas, rng, 0);
+    if (second_random_idx >= first_random_idx) {
+      second_random_idx++;
+    }
```
What's the purpose of this block?

AleCarminati replied:
From line 34 to 40 we have to sample two datapoints' indexes.
For the first index, we sample from 0 to n_data-1.
The second index will be an index from 0 to n_data-1, with the exception of the first index. Therefore, we sample from 0 to (n_data-1)-1, then, if the sampled value is greater than the first index, we increase it by one. In this way, it is like sampling from 0 to n_data-1, leaving a "hole" where before there was the first index.

brunoguindani commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,393 @@
+          unique_values[allocations[random_idxs[clust_idx]]]->get_data_idx();
+    }
+    set_to_cycle.erase(random_idxs[clust_idx]);
+    auto it = set_to_cycle.cbegin();
+    const auto end = set_to_cycle.cend();
+
+    for (; it != end; ++it) {
```
If you don't need the two iterators outside of the for loop, please declare them in the `()`

brunoguindani commented on src/algorithms/split_and_merge_algorithm.cc:
```diff
@@ -0,0 +1,393 @@
+  return log_res_prod;
+}
+
+void SplitAndMergeAlgorithm::full_gibbs_sampling() {
```
This code looks very similar to something found in Neal2, but you already discussed this if I recall correctly. If so, please place a "TODO: identical to Neal2" so that we remember to handle this in the future

brunoguindani reviewed (commented):
Mostly minor remarks about comments and formatting from my side!

AleCarminati commented:
With the build test on commit [1b3d1b7](https://github.com/bayesmix-dev/bayesmix/pull/116/commits/1b3d1b777cb442938d6d8469428d03eed9f2c1bc) we noted that our fork was not up-to-date with the changes in the signatures of `get_mass_new_cluster` and `get_mass_existing_cluster`. Therefore, we fetched upstream our fork and we modified the calls to these functions. Now it should work correctly.

brunoguindani reviewed (approved):
We good! @mberaha, please merge at will if there isn't anything else to do

mberaha commented:
Way to go! Guys, before merging, I would like to ask you one last thing: create a python notebook in `python/notebooks` that you call something like "split_merge_benchmarking.ipynb" where you put all the experiments you have and the comparison with Neal2/3/8. 
You can either use the command line via an executable file to run all the experiments or use the `bayesmixpy` interface (it's just a wrapper around the command line, nothing fancy about it) and report the plots etc...

Let me know when you have something ready :)

AleCarminati commented:
We just added the notebook, as you suggested. It contains the tests that we ran to obtain the numerical values and the graphs for the final presentation and the report.
It is composed of two sections: in the first one we test the algorithm on galaxy and faithful, in the second one we test the algorithm with high dimensional synthetic datasets. As you suggested, we tried different combinations of parameters to check if there were some advantages in using Split and Merge. We found a combination (the one that is written in the notebook) in which Split and Merge clearly outclasses Neal3: with dimensions between 6 and 10, Split and Merge recognizes correctly the two clusters, while Neal3 in some cases is not able to find two clusters and in other cases requires thousands of iterations to find them.

mberaha commented:
Impressive job guys! I'm really glad to merge this PR to master!

mberaha commented:
We need more developers like these ones!!

https://www.youtube.com/watch?v=EMldOiiG1Ko

AleCarminati commented:
Thank you for the patience and all the suggestions!
